compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala

```diff
@@ -1184,15 +1184,17 @@ class TreeUnpickler(reader: TastyReader,
             case SELECTin =>
               var sname = readName()
               val qual = readTerm()
-              val owner = readType()
-              def select(name: Name, denot: Denotation) =
-                val prefix = ctx.typeAssigner.maybeSkolemizePrefix(qual.tpe.widenIfUnstable, name)
-                makeSelect(qual, name, denot.asSeenFrom(prefix))
-              sname match
-                case SignedName(name, sig, target) =>
-                  select(name, owner.decl(name).atSignature(sig, target))
-                case name =>
-                  select(name, owner.decl(name))
+              val owner = readType().typeSymbol
+              val SignedName(name, sig, target) = sname: @unchecked // only methods with params use SELECTin
+              val qualType = qual.tpe.widenIfUnstable
+              val prefix = ctx.typeAssigner.maybeSkolemizePrefix(qualType, name)
+              def matchOverload(denot: SingleDenotation) =
+                val sym = denot.symbol
+                sym.exists
+                && sym.owner == owner
+                && sym.targetName == target
+                && sym.signature == sig
+              makeSelect(qual, name, qualType.findMember(name, prefix).filterWithPredicate(matchOverload))
```

Comment:
Thinking about it more I don't think we can change to using findMember on the qualifier here: the problem is that if an override of the symbol is added, findMember will pick up this override which will have a different symbol and signature. We could look through the overridden symbols to find a match but that would be a bit expensive. So I think we should go back to the previous version and just use SELECT for refinements (and maybe completely disallow refinements of overloaded methods if that isn't already the case).

By the way, we absolutely need test cases to ensure we can add overloads and overrides without breaking tasty compatibility. Unfortunately we don't have a good infrastructure to do that currently, last time I needed to do this I set up an sbt scripted test where one project is manually added to the classpath of the other project to bypass incremental compilation: https://github.com/lampepfl/dotty/commit/66e5ad3482317098ac3bcfb76cabe43c06e600cc

Reply:
To be efficient here, should we then record when a denotation has been selected from a refinement?

Reply:
Sorry, to be efficient at what?

Reply:
I have made a test where we have "library" `A`:
```scala
// A.scala
package a

object A {

  trait Box0[A] {
    def append(a: A): this.type = this
    def append(a: A*): this.type = this
  }

  trait Box[A] extends Box0[A]

  val box: Box[Seq[Double]] = new Box[Seq[Double]] {}

}
```
and the "client" code `B`:
```scala
// B.scala
import a.*

object B {
  A.box.append(Seq.empty[Double])
}
```

if we compile both, and then recompile A like so
```scala
// A-changes.scala
package a

object A {

  trait Box0[A] {
    def append(a: A*): this.type = this
  }

  trait Box[A] extends Box0[A] {
    def append(a: A): this.type = this
  }

  val box: Box[Seq[Double]] = new Box[Seq[Double]] {}

}
```
Then with the present changes in 013e08a trying to compile `B.tasty` against the new `A.tasty` will fail.
```scala
-- Error: b/B.scala:4:2 --------------------------------------------------------
4 |  A.box.append(Seq.empty[Double])
  |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |undefined: a.A.box.append # -1: TermRef(TermRef(TermRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object a),A),val box),append) at readTasty
```

Reply:
although in this case as we have removed the overload we chose before I guess this is an expected error?

Reply:
@smarter previously you said there was no way to detect if a denotation comes from a refinement, so you must do an expensive operation to detect if it came from a refinement, if we want to use SELECT for refinements.